python/cudf_polars/cudf_polars/dsl/traversal.py

```diff
@@ -0,0 +1,191 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+# SPDX-License-Identifier: Apache-2.0
+
+"""Traversal and visitor utilities for nodes."""
+
+from __future__ import annotations
+
+from collections.abc import Hashable
+from typing import TYPE_CHECKING, Any, Generic, TypeVar, overload
+
+if TYPE_CHECKING:
+    from collections.abc import Callable, Generator, Mapping, MutableMapping
+
+    from cudf_polars.dsl import expr, ir
+    from cudf_polars.dsl.nodebase import Node
+    from cudf_polars.typing import ExprTransformer, GenericTransformer, IRTransformer
+
+
+__all__: list[str] = [
+    "traversal",
+    "reuse_if_unchanged",
+    "make_recursive",
+    "CachingVisitor",
+]
+
+
+def traversal(node: Node) -> Generator[Node, None, None]:
+    """
+    Pre-order traversal of nodes in an expression.
+
+    Parameters
+    ----------
+    node
+        Root of expression to traverse.
+
+    Yields
+    ------
+    Unique nodes in the expression, parent before child, children
+    in-order from left to right.
+    """
+    seen = {node}
+    lifo = [node]
+
+    while lifo:
+        node = lifo.pop()
+        yield node
+        for child in reversed(node.children):
+            if child not in seen:
+                seen.add(child)
+                lifo.append(child)
+
+
+# reuse_if_unchanged can either be applied to expressions, in which
+# case we need an ExprTransformer...
+@overload
+def reuse_if_unchanged(node: expr.Expr, fn: ExprTransformer) -> expr.Expr: ...
+
+
+# .. or to plan nodes (IR), in which case we need an IRTransformer
+@overload
+def reuse_if_unchanged(node: ir.IR, fn: IRTransformer) -> ir.IR: ...
+
+
+def reuse_if_unchanged(node, fn):
+    """
+    Recipe for transforming nodes that returns the old object if unchanged.
+
+    Parameters
+    ----------
+    node
+         Node to recurse on
+    fn
+         Function to transform children
+
+    Notes
+    -----
+    This can be used as a generic "base case" handler when
+    writing transforms that take nodes and produce new nodes.
+
+    Returns
+    -------
+    Existing node `e` if transformed children are unchanged, otherwise
+    reconstructed node with new children.
+    """
+    new_children = [fn(c) for c in node.children]
+    if all(new == old for new, old in zip(new_children, node.children, strict=True)):
+        return node
```

Comment:
Would it make sense to redefine a transformer as returning a pair where the second element is a boolean indicating whether any modifications were made? It would avoid needing checks like this, but it might clutter the code too much to be worthwhile.

Reply:
It's kind of ugly because now everyone has to do the zip-transpose dance when recursing:

```
def reuse_if_unchanged(node: Node, rec: Transformer) -> tuple[Node, bool]:
    children, changes = zip(*map(rec, node.children))
    if any(changes):
        return node.reconstruct(children), True
    return node, False
```

Which is not too bad hidden away, but you need to handle that again, every time.

Because of the hash-caching and CSE-elimination in `Node` for equality, this check for equality is actually fast since as you come back up the recursion you only ever end up going one-level down in checks, since you fail fast and if objects are equal after the recursion, they will have object-identical children.